gnupg-1.4.23/doc/gpgv.1
.\" Created from Texinfo source by yat2m 1.0
.TH GPGV 1 2024-11-17 "GnuPG 1.4.23" "GNU Privacy Guard 1.4"
.SH NAME
.B gpgv
\- Verify OpenPGP signatures
.SH SYNOPSIS
.B  gpgv
.RI [ options ]
.I signed_files




.SH DESCRIPTION
\fBgpgv\fR is an OpenPGP signature verification tool.

This program is actually a stripped-down version of \fBgpg\fR which is
only able to check signatures. It is somewhat smaller than the fully-blown
\fBgpg\fR and uses a different (and simpler) way to check that
the public keys used to make the signature are valid. There are
no configuration files and only a few options are implemented.

\fBgpgv\fR assumes that all keys in the keyring are trustworthy.
That does also mean that it does not check for expired or revoked
keys.

By default a keyring named \(oq\fItrustedkeys.gpg\fR\(cq is used.  This
default keyring is assumed to be in the home directory of GnuPG,
either the default home directory or the one set by an option or an
environment variable.  The option \fB--keyring\fR may be used to
specify a different keyring or even multiple keyrings.



.SH RETURN VALUE

The program returns 0 if everything is fine, 1 if at least
one signature was bad, and other error codes for fatal errors.

.SH OPTIONS
\fBgpgv\fR recognizes these options:


.TP
.B  --verbose
.TP
.B  -v
Gives more information during processing. If used
twice, the input data is listed in detail.

.TP
.B  --quiet
.TP
.B  -q
Try to be as quiet as possible.

.TP
.B  --keyring \fIfile\fR
Add \fIfile\fR to the list of keyrings.
If \fIfile\fR begins with a tilde and a slash, these
are replaced by the HOME directory. If the filename
does not contain a slash, it is assumed to be in the
home-directory ("~/.gnupg" if --homedir is not used).

.TP
.B  --status-fd \fIn\fR
Write special status strings to the file descriptor \fIn\fR.  See the
file DETAILS in the documentation for a listing of them.

.TP
.B  --logger-fd \fBn\fR
Write log output to file descriptor \fBn\fR and not to stderr.

.TP
.B  --ignore-time-conflict
GnuPG normally checks that the timestamps associated with keys and
signatures have plausible values. However, sometimes a signature seems to
be older than the key due to clock problems. This option turns these
checks into warnings.

.TP
.B  --homedir \fIdir\fR
Set the name of the home directory to \fIdir\fR. If this option is not
used, the home directory defaults to \(oq\fI~/.gnupg\fR\(cq.  It is only
recognized when given on the command line.  It also overrides any home
directory stated through the environment variable \(oq\fIGNUPGHOME\fR\(cq or
(on Windows systems) by means of the Registry entry
\fIHKCU\\Software\\GNU\\GnuPG:HomeDir\fR.

On Windows systems it is possible to install GnuPG as a portable
application.  In this case only this command line option is
considered, all other ways to set a home directory are ignored.

To install GnuPG as a portable application under Windows, create an
empty file name \(oq\fIgpgconf.ctl\fR\(cq in the same directory as the tool
\(oq\fIgpgconf.exe\fR\(cq.  The root of the installation is than that
directory; or, if \(oq\fIgpgconf.exe\fR\(cq has been installed directly below
a directory named \(oq\fIbin\fR\(cq, its parent directory.  You also need to
make sure that the following directories exist and are writable:
\(oq\fIROOT/home\fR\(cq for the GnuPG home and \(oq\fIROOT/var/cache/gnupg\fR\(cq
for internal cache files.

.TP
.B  --weak-digest \fBname\fR
Treat the specified digest algorithm as weak.  Signatures made over
weak digests algorithms are normally rejected. This option can be
supplied multiple times if multiple algorithms should be considered
weak.  MD5 is always considered weak, and does not need to be listed
explicitly.


.SH EXAMPLES


.TP
.B  gpgv \fBpgpfile\fR
.TP
.B  gpgv \fBsigfile\fR [\fBdatafile\fR]
Verify the signature of the file. The second form is used for detached
signatures, where \fBsigfile\fR is the detached signature (either
ASCII-armored or binary) and \fBdatafile\fR contains the signed data;
if \fBdatafile\fR is "-" the signed data is expected on
\fBstdin\fR; if \fBdatafile\fR is not given the name of the file
holding the signed data is constructed by cutting off the extension
(".asc", ".sig" or ".sign") from \fBsigfile\fR.


.SH FILES


.TP
.B  ~/.gnupg/trustedkeys.gpg
The default keyring with the allowed keys.


.SH ENVIRONMENT


.TP
.B  HOME
Used to locate the default home directory.

.TP
.B  GNUPGHOME
If set directory used instead of "~/.gnupg".


.SH SEE ALSO
\fBgpg2\fR(1)

The full documentation for this tool is maintained as a Texinfo manual.
If GnuPG and the info program are properly installed at your site, the
command

.RS 2
.nf
info gnupg
.fi
.RE

should give you access to the complete manual including a menu structure
and an index.



gnupg-2.4.6/ChangeLog
2024-10-29  Werner Koch  <wk@gnupg.org>

	Release 2.4.6.
	+ commit 5340576c08309c9a22b1469464052c36d511d0d3


2024-10-29  Ingo Klöcker  <dev@ingo-kloecker.de>

	gpg: Fix --quick-set-expire for V5 subkey fingerprints.
	+ commit f355f3f3818211e4c977c6c2e60e0d2f8c7be945
	* g10/keyedit.c (keyedit_quick_set_expire): Use actual size of
	fingerprint.

2024-10-29  NIIBE Yutaka  <gniibe@fsij.org>

	common: Fix a race condition in creating socketdir.
	+ commit 2ae017a25fe800c9bc6fc1090179ec987bb8ba8d
	* common/homedir.c (_gnupg_socketdir_internal): Check return code
	of gnupg_mkdir and handle the case of GPG_ERR_EEXIST.

2024-10-15  Werner Koch  <wk@gnupg.org>

	speedo: Enable additional runtime protections on Windows.
	+ commit afe87ffc08e14317f4ef5bbe2940d07203a43808
	* build-aux/speedo.mk (speedo_w32_cflags): Remove -mms-bitfields
	because it is for a long time the gcc default.  Enable control flow
	protection.

2024-10-07  Werner Koch  <wk@gnupg.org>

	gpg: Emit status error for an invalid ADSK.
	+ commit a8b503c42bd4a773c6a9f88f553138d8092eec4d
	* g10/keygen.c (prepare_adsk): Emit status error.

2024-09-25  Werner Koch  <wk@gnupg.org>

	gpg: Exclude expired trusted keys from the key validation process.
	+ commit 8b4ad616de6cc1bc4c76885a41b95f2edfdf93a0
	* g10/trustdb.c (copy_key_item): New.
	(validate_keys): Use a stripped down UTK list w/o expired keys.

	gpg: Validate the trustdb after the import of a trusted key.
	+ commit ceec31751cc0683cec96d0466dfc2b9ef0dc0913
	* g10/import.c (import_one_real): Rename non_self to non_self_or_utk.
	If not set after chk_self_sigs check whether the imported key is an
	ultimately trusted key.

	gpg: Remove useless variable in validate_keys.
	+ commit ce54266d66703e1a0227d6e16e0c9cd1a422bc9e
	* g10/trustdb.c (store_validation_status): Remove arg  'stored'.
	(validate_keys): Remove keyhashtable 'stored' which was never used.

2024-09-20  Werner Koch  <wk@gnupg.org>

	w32: Fix last commit to build on Windows.
	+ commit 03459618c5f80fa05d1a88074f8a72d37535f804
	* scd/app.c (struct mrsw_lock): Move notify_watchers out of the system
	specific condition.

2024-09-20  NIIBE Yutaka  <gniibe@fsij.org>

	scd: Fix DEVINFO, allowing no clients which watch the change.
	+ commit c98385d311ca37e1863d0e42ebf7bbc6b68efe35
	* scd/app.c [POSIX] (struct mrsw_lock): Add notify_watchers.
	(card_list_signal): Only when watchers wait, kick by write(2).
	(card_list_wait): Increment/decrement notify_watchers field.

2024-09-19  Werner Koch  <wk@gnupg.org>

	speedo: Make use of wget more robust.
	+ commit 94a7a674737b9cda196252c491af0a9ad023f3de
	* build-aux/getswdb.sh: Add option --wgetopt.
	* build-aux/speedo.mk (WGETOPT): New.
	(getswdb_options): Pass to getswdb.
	(unpack): Use wget with new options.

	doc: Add support for generating HTML versions of the man pages.
	+ commit 35d80ebd7876fa3071c1a88c7ad0b9a0022dc418
	* doc/Makefile.am (yat2m-stamp): Also call yat2m with --html options.
	* doc/yat2m.c (main): Add dummy options.

2024-09-19  NIIBE Yutaka  <gniibe@fsij.org>

	kbx: Fix a race condition on DATABASE_HD.
	+ commit a698adbb533fac12277aaf319ab08d0e7f027261
	* kbx/backend-sqlite.c (create_or_open_database): Protect
	the access to DATABASE_HD.

	scd: Fix DEVINFO to allow multiple clients.
	+ commit 32476f870d629dd4408e7a7fea6e0c867b8fd84a
	* scd/app.c (initialize_module_command): Use O_NONBLOCK for pipe.

2024-09-19  Werner Koch  <wk@gnupg.org>

	build: Also cleanup generated html file in a make distcheck.
	+ commit 685acf650a68871d8effeaa1c2ca01f2c4752266
	* doc/Makefile.am (myman_pages): Add gpg and gpgv.
	(USE_GPG2_HACK): Remove conditional.
	(myhtmlman_pages): New.
	(DISTCLEANFILES): Add html pages.

2024-09-19  NIIBE Yutaka  <gniibe@fsij.org>

	gpg: Fix getting key by IPGP.
	+ commit a17584d000cfae787fc187ed19c0d9cb94f3ca48
	* g10/call-dirmngr.c (gpg_dirmngr_dns_cert): Check if DATA for key.

2024-09-19  Werner Koch  <wk@gnupg.org>

	gpg: Improve detection of input data read errors.
	+ commit 178c3fe62c251c762152a7dc883f009d41f08244
	* g10/build-packet.c (do_plaintext): Better error checking for
	iobuf_copy.

	gpg: Make --no-literal work again for -c and --store.
	+ commit e16728f3d63dbadd176e6390514c90f8a11f9b30
	* g10/dearmor.c (dearmor_file): Check for errors of iobuf_copy.
	(enarmor_file): Ditto.
	* g10/encrypt.c (encrypt_simple): Fix error check of iobuf_copy
	(encrypt_crypt): Use iobuf_copy.

	gpg: remove workaround for Libgcrypt < 1.8.6.
	+ commit 0f0c59d6ff9dda23f04dacdcfbe937b7a6e5625b
	* g10/free-packet.c (is_mpi_copy_broken): Remove.

	gpg: Avoid wrong decryption_failed for signed+OCB msg w/o pubkey.
	+ commit e8598390becaedb54c4eba31d5e3e047a091d45f
	* g10/decrypt-data.c (struct decode_filter_context_s): Add flag
	checktag_failed.
	(aead_checktag): Set flag.
	(decrypt_data): Initially clear that flag and check the flag after the
	decryption.
	* g10/mainproc.c (proc_encrypted): Revert the log_get_errorcount based
	check.

2024-09-17  Werner Koch  <wk@gnupg.org>

	agent: Fix detection of the trustflag de-vs.
	+ commit 6ff13380a2e3226f675740ee66c8d56fa6f5d110
	* agent/trustlist.c (read_one_trustfile): Fix comparison.

2024-09-17  NIIBE Yutaka  <gniibe@fsij.org>

	common:w32: Don't expose unused functions.
	+ commit 730593affa91182ccd57d9618dd650fe170ebe1c
	* common/exechelp.h [HAVE_W32_SYSTEM] (get_max_fds): Don't expose.
	(close_all_fds, get_all_open_fds): Likewise.
	* common/exechelp-w32.c: Don't expose unused functions.

2024-09-13  Werner Koch  <wk@gnupg.org>

	gpgsm: New option --assert-signer.
	+ commit 54e06273c0dc9482941481e304d919e410fa2832
	* sm/gpgsm.c (oAssertSigner, oNoop): New.
	(opts): Add option --assert-signer.
	(assert_signer_true): New var.
	(main): Set new option.
	(gpgsm_exit): Handle assert_signer_true.
	* sm/gpgsm.h (opt): Add field assert_signer_list.
	* sm/verify.c (is_x509_fingerprint): New.
	(check_assert_signer_list): New.
	(gpgsm_verify): Handle option.

2024-09-03  Werner Koch  <wk@gnupg.org>


gnupg-1.4.23/g10/gpgv.c
/* gpgv.c - The GnuPG signature verify utility
 * Copyright (C) 1998, 1999, 2000, 2001, 2002, 2005, 2006,
 *               2009, 2012 Free Software Foundation, Inc.
 *
 * This file is part of GnuPG.
 *
 * GnuPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * GnuPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#ifdef HAVE_DOSISH_SYSTEM
#include <fcntl.h> /* for setmode() */
#endif
#ifdef HAVE_LIBREADLINE
#include <stdio.h>
#include <readline/readline.h>
#endif

#define INCLUDED_BY_MAIN_MODULE 1
#include "packet.h"
#include "iobuf.h"
#include "memory.h"
#include "util.h"
#include "main.h"
#include "options.h"
#include "keydb.h"
#include "trustdb.h"
#include "mpi.h"
#include "cipher.h"
#include "filter.h"
#include "ttyio.h"
#include "i18n.h"
#include "status.h"
#include "cardglue.h"


enum cmd_and_opt_values { aNull = 0,
    oQuiet	  = 'q',
    oVerbose	  = 'v',
    oBatch	  = 500,
    oKeyring,
    oIgnoreTimeConflict,
    oStatusFD,
    oLoggerFD,
    oHomedir,
    oWeakDigest,
aTest };


static ARGPARSE_OPTS opts[] = {

    { 301, NULL, 0, N_("@\nOptions:\n ") },

    { oVerbose, "verbose",   0, N_("verbose") },
    { oQuiet,	"quiet",     0, N_("be somewhat more quiet") },
    { oKeyring, "keyring"   ,2, N_("take the keys from this keyring")},
    { oIgnoreTimeConflict, "ignore-time-conflict", 0,
                           N_("make timestamp conflicts only a warning") },
    { oStatusFD, "status-fd" ,1, N_("|FD|write status info to this FD") },
    { oLoggerFD, "logger-fd",1, "@" },
    { oHomedir, "homedir", 2, "@" },   /* defaults to "~/.gnupg" */
    { oWeakDigest, "weak-digest", 2, "@" },   /* defaults to "~/.gnupg" */

{0} };



int g10_errors_seen = 0;

const char *
strusage( int level )
{
    const char *p;
    switch( level ) {
      case 11: p = "gpgv (GnuPG)";
	break;
      case 13: p = VERSION; break;
      case 17: p = PRINTABLE_OS_NAME; break;
      case 19: p =
	    _("Please report bugs to <gnupg-bugs@gnu.org>.\n");
	break;
      case 1:
      case 40:	p =
	    _("Usage: gpgv [options] [files] (-h for help)");
	break;
      case 41:	p =
	    _("Syntax: gpgv [options] [files]\n"
	      "Check signatures against known trusted keys\n");
	break;

      default:	p = default_strusage(level);
    }
    return p;
}




static void
i18n_init(void)
{
#ifdef USE_SIMPLE_GETTEXT
    set_gettext_file (PACKAGE, "Software\\GNU\\GnuPG");
#else
#ifdef ENABLE_NLS
    setlocale( LC_ALL, "" );
    bindtextdomain (PACKAGE, LOCALEDIR);
    textdomain( PACKAGE );
#endif
#endif
}


int
main( int argc, char **argv )
{
    ARGPARSE_ARGS pargs;
    int rc=0;
    STRLIST sl;
    STRLIST nrings=NULL;
    unsigned configlineno;

    log_set_name("gpgv");
    init_signals();
    i18n_init();
    opt.command_fd = -1; /* no command fd */
    opt.pgp2_workarounds = 1;
    opt.keyserver_options.options|=KEYSERVER_AUTO_KEY_RETRIEVE;
    opt.trust_model = TM_ALWAYS;
    opt.no_sig_cache = 1;
    opt.flags.require_cross_cert = 1;
    opt.batch = 1;
    opt.weak_digests = NULL;

    opt.homedir = default_homedir ();

    tty_no_terminal(1);
    tty_batchmode(1);
    dotlock_disable ();

    set_native_charset (NULL); /* Try to auto set the character set */
    additional_weak_digest("MD5");

    pargs.argc = &argc;
    pargs.argv = &argv;
    pargs.flags=  1;  /* do not remove the args */
    while( optfile_parse( NULL, NULL, &configlineno, &pargs, opts) ) {
	switch( pargs.r_opt ) {
	  case oQuiet: opt.quiet = 1; break;
          case oVerbose: g10_opt_verbose++;
		  opt.verbose++; opt.list_sigs=1; break;
          case oKeyring: append_to_strlist( &nrings, pargs.r.ret_str); break;
	  case oStatusFD: set_status_fd( pargs.r.ret_int ); break;
	  case oLoggerFD: log_set_logfile( NULL, pargs.r.ret_int ); break;
	  case oHomedir: opt.homedir = pargs.r.ret_str; break;
	  case oWeakDigest: additional_weak_digest(pargs.r.ret_str); break;
	  case oIgnoreTimeConflict: opt.ignore_time_conflict = 1; break;
	  default : pargs.err = 2; break;
	}
    }

    if( log_get_errorcount(0) ){

	g10_exit(2);}

    g10_opt_homedir = opt.homedir;

    if( opt.verbose > 1 )
	set_packet_list_mode(1);

    if( !nrings )  /* no keyring given: use default one */
        keydb_add_resource ("trustedkeys" EXTSEP_S "gpg", 8, 0);
    for(sl = nrings; sl; sl = sl->next )
        keydb_add_resource (sl->d, 8, 0 );

    FREE_STRLIST(nrings);

    if( (rc = verify_signatures( argc, argv ) ))
        log_error("verify signatures failed: %s\n", g10_errstr(rc) );

    /* cleanup */
    g10_exit(0);
    return 8; /*NEVER REACHED*/
}


void
g10_exit( int rc )
{
    rc = rc? rc : log_get_errorcount(0)? 2 :
			g10_errors_seen? 1 : 0;
    exit(rc );
}


/* Stub:
 * We have to override the trustcheck from pkclist.c becuase
 * this utility assumes that all keys in the keyring are trustworthy
 */
int
check_signatures_trust( PKT_signature *sig )
{
    return 0;
}

void
read_trust_options(byte *trust_model,ulong *created,ulong *nextcheck,
		   byte *marginals,byte *completes,byte *cert_depth,
		   byte *min_cert_level) {}

/* Stub:
 * We don't have the trustdb , so we have to provide some stub functions
 * instead
 */

int
cache_disabled_value(PKT_public_key *pk)
{
  return 0;
}

void
check_trustdb_stale(void) {}

int
get_validity_info (PKT_public_key *pk, PKT_user_id *uid)
{
    return '?';
}

unsigned int
get_validity (PKT_public_key *pk, PKT_user_id *uid)
{
  return 0;
}

const char *
trust_value_to_string (unsigned int value)
{
  return "err";
}

const char *
uid_trust_string_fixed(PKT_public_key *key,PKT_user_id *uid)
{
  return "err";
}

int
get_ownertrust_info (PKT_public_key *pk)
{
    return '?';
}

unsigned int
get_ownertrust (PKT_public_key *pk)
{
    return TRUST_UNKNOWN;
}


/* Stubs:
 * Because we only work with trusted keys, it does not make sense to
 * get them from a keyserver
 */

struct keyserver_spec *
keyserver_match(struct keyserver_spec *spec) { return NULL; }

int
keyserver_import_keyid( u32 *keyid, void *dummy )
{
    return -1;
}

int
keyserver_import_cert(const char *name) { return -1; }

int
keyserver_import_pka(const char *name,unsigned char *fpr) { return -1; }

int
keyserver_import_name(const char *name,struct keyserver_spec *spec)
{
  return -1;
}

int
keyserver_import_ldap(const char *name) { return -1; }

/* Stub:
 * No encryption here but mainproc links to these functions.
 */
int
get_session_key( PKT_pubkey_enc *k, DEK *dek )
{
    return G10ERR_GENERAL;
}
/* Stub: */
int
get_override_session_key( DEK *dek, const char *string )
{
    return G10ERR_GENERAL;
}
/* Stub: */
int
decrypt_data( void *procctx, PKT_encrypted *ed, DEK *dek )
{
    return G10ERR_GENERAL;
}


/* Stub:
 * No interactive commnds, so we don't need the helptexts
 */
void
display_online_help( const char *keyword )
{
}

/* Stub:
 * We don't use secret keys, but getkey.c links to this
 */
int
check_secret_key( PKT_secret_key *sk, int n )
{
    return G10ERR_GENERAL;
}

/* Stub:
 * No secret key, so no passphrase needed
 */
DEK *
passphrase_to_dek( u32 *keyid, int pubkey_algo,
		   int cipher_algo, STRING2KEY *s2k, int mode,
                   const char *tmp, int *canceled)
{
  if (canceled)
    *canceled = 0;
  return NULL;
}

struct keyserver_spec *parse_preferred_keyserver(PKT_signature *sig) {return NULL;}
struct keyserver_spec *parse_keyserver_uri(const char *uri,int require_scheme,
                                           const char *configname,
                                           unsigned int configlineno)
{
  return NULL;
}

void free_keyserver_spec(struct keyserver_spec *keyserver) {}

/* Stubs to avoid linking to photoid.c */
void show_photos(const struct user_attribute *attrs,int count,PKT_public_key *pk) {}
int parse_image_header(const struct user_attribute *attr,byte *type,u32 *len) {return 0;}
char *image_type_to_string(byte type,int string) {return NULL;}

#ifdef ENABLE_CARD_SUPPORT
int agent_scd_getattr (const char *name, struct agent_card_info_s *info) {return 0;}
#endif /* ENABLE_CARD_SUPPORT */

/* Stubs to void linking to ../cipher/cipher.c */
int string_to_cipher_algo( const char *string ) { return 0; }
const char *cipher_algo_to_string( int algo ) { return "?";}
void disable_cipher_algo( int algo ) {}
int check_cipher_algo( int algo ) { return -1;}
unsigned int cipher_get_keylen( int algo ) { return 0; }
unsigned int cipher_get_blocksize( int algo ) {return 0;}
CIPHER_HANDLE cipher_open( int algo, int mode, int secure ) { return NULL;}
void cipher_close( CIPHER_HANDLE c ) {}
int cipher_setkey( CIPHER_HANDLE c, byte *key, unsigned keylen ) { return -1;}
void cipher_setiv( CIPHER_HANDLE c, const byte *iv, unsigned ivlen ){}
void cipher_encrypt( CIPHER_HANDLE c, byte *outbuf,
                     byte *inbuf, unsigned nbytes ) {}
void cipher_decrypt( CIPHER_HANDLE c, byte *outbuf,
                     byte *inbuf, unsigned nbytes ) {}
void cipher_sync( CIPHER_HANDLE c ) {}

/* Stubs to avoid linking to ../cipher/random.c */
void random_dump_stats(void) {}
int quick_random_gen( int onoff ) { return -1;}
void randomize_buffer( byte *buffer, size_t length, int level ) {}
void randomize_mpi (MPI mpi, size_t nbits, int level) {}
int random_is_faked() { return -1;}
byte *get_random_bits( size_t nbits, int level, int secure ) { return NULL;}
void set_random_seed_file( const char *name ) {}
void update_random_seed_file() {}
void fast_random_poll() {}

/* Stubs to avoid linking of ../cipher/primegen.c */
void register_primegen_progress ( void (*cb)( void *, int), void *cb_data ) {}
MPI generate_secret_prime( unsigned  nbits ) { return NULL;}
MPI generate_public_prime( unsigned  nbits ) { return NULL;}
MPI generate_elg_prime( int mode, unsigned pbits, unsigned qbits,
                        MPI g, MPI **ret_factors ) { return NULL;}

/* Do not link to ../cipher/rndlinux.c */
void rndlinux_constructor(void) {}


/* Stubs to avoid linking to ../util/ttyio.c */
int tty_batchmode( int onoff ) { return 0; }
void tty_printf( const char *fmt, ... ) { }
void tty_fprintf (FILE *fp, const char *fmt, ... ) { }
void tty_print_string( const byte *p, size_t n ) { }
void tty_print_utf8_string( const byte *p, size_t n ) {}
void tty_print_utf8_string2( const byte *p, size_t n, size_t max_n ) {}
char *tty_get( const char *prompt ) { return NULL;}
char *tty_get_hidden( const char *prompt ) {return NULL; }
void tty_kill_prompt(void) {}
int tty_get_answer_is_yes( const char *prompt ) {return 0;}
int tty_no_terminal(int onoff) {return 0;}
void tty_cleanup_after_signal (void) {}
#ifdef __VMS
FILE *ttyfp_is (void) { return stderr; }
void init_ttyfp (void) { }
#endif /*__VMS*/
#ifdef HAVE_LIBREADLINE
void tty_enable_completion(rl_completion_func_t *completer) {}
void tty_disable_completion(void) {}
void rl_cleanup_after_signal (void) {}
void rl_free_line_state (void) {}
#endif

/* We do not do any locking, so use these stubs here */
void dotlock_disable(void) {}
dotlock_t dotlock_create (const char *file_to_lock, unsigned int flags)
{ return NULL; }
void dotlock_destroy (dotlock_t h, int reclaim) {}
int dotlock_take (dotlock_t h, long timeout) { return 0;}
int dotlock_release (dotlock_t h) {return 0;}
void dotlock_remove_lockfiles (int reclaim) {}
